DeerajTheepshi/processing-android
# [FEAT]Emulator Creator, Emulator Configuration, Emulator flow

# What :
- This PR features the all new emulator creation flow
- It also features Run Configuration menu item , from where users can select their SDK Platform, select their Emulator (if they have more than 1) .
- Users can also delete or create a new AVD from the Run Configuration window

# Why :
- To give users the choice of running their sketch on different emulator screens. 
- To give users the choice to name, create, configure and delete their emulators. 
- To allow users to make use of different platforms that they download from the SDK Updater .

# How :
- New UI is handled by two classes : CreateAVD.java (The entire flow) and RunConfiguration.java (The menu item) : These two classes are pure UI classes and handle only the corresponding UI.
- New methods to enable are created in AVD.java - This class has some unwanted code that will be removed in further commits.
- The image downloader class - SysimageDownloader.java -  This class was modified to download the selected image, rather than the always default image.
- Emulatorcontroller.java - This class was modified to start the emulator from the given name, rather than starting the default one.

# Note :
- Swing worker has not yet been implemented. (Following commits will do it) **(Completed with d347fd196fd75e4d713dd3cac44d350e643f4dcf)**
- Unwanted code to be remove. **(Completed with dda764d6d914f74ca80f8f4f173a35351df052f7)**
- Add listener to table, so that delete button is enabled only after clicking an item in the table
- Handle back button in create avd windows. **(Completed with d347fd196fd75e4d713dd3cac44d350e643f4dcf)**

# Commit info : 
- Device selector added using : e1165a2c85bf618dcb365005f991987a50a15d61
- What does this do ? Now devices can be selected from the run configuration , just like how you select emulators

# Testing Guidelines : 
- Remove everything related to android emulator from the preferences file. 
- Open processing and select 1.Android >> Run Configurations where you can create new emulators or choose existing ones. 2. You can directly click run on emulator to prompt for new emulator creation.
- **Option 1**
    - Select your Target SDK. A list of all your available SDKs will be shown.
    - Hit create button to go through the process of creating a new emulator.
    - After new emulator is created, you can delete or select and run your sketch on it.
    - Choose the devices radio button to select your device. If no device is found the button will automatically shift back to emulator radio button.
    - During delete option, the UX goes into a minor waiting state, during which message is displayed on the console and after deleting , the table is again refreshed.
- **Option 2**
     - This walks you through a new emulator creation process
     - First fill in your device name , and pick a emulator (phone or wear) , and then a phone device of your choice.
     - After this, the UX goes into loading state, while it is querying for system images in the background. Once query is completed, it presents you with a table of system images.
     -  The already installed system images are shown on top of the table, these can be selected and next button be pressed.
     - If the system image you want is not installed, then click the Download button to download the System-image. Once download is complete via a simple progress bar, , reselct the image and hit next button.
     - This is the final step, confirm your emulator. Emulator creation might take some time. After which the build process will continue. 
      - **Note : ** in case of create button in Run configuration window mention in option 1, the same steps follow.
- **NOTE: ** In either of the cases, the newly created emulator or the selected emulator will be saved and used for future sketch runs. 
- Test deleting the Emulator : Try deleting the previously selected emulator via Run Configuration or through command line using avdmanager. Open processing and choose run . This time processing will detect that the saved emulator is no longer present and will stop the run process and inform you to change the emulator in run configuration window.
